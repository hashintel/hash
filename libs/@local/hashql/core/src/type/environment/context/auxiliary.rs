use crate::r#type::{
    collection::FastHashMap,
    kind::generic_argument::{GenericArgumentData, GenericArgumentId},
};

#[derive(Debug)]
pub struct AuxiliaryData {
    pub arguments: FastHashMap<GenericArgumentId, GenericArgumentData>,
}

impl AuxiliaryData {
    #[must_use]
    pub fn new() -> Self {
        Self {
            arguments: FastHashMap::default(),
        }
    }
}

impl Default for AuxiliaryData {
    fn default() -> Self {
        Self::new()
    }
}
