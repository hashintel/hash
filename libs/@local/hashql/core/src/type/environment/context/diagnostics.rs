use crate::r#type::error::TypeCheckDiagnostic;

/// A collection of type checking diagnostics produced during compilation.
///
/// This struct maintains a list of [`TypeCheckDiagnostic`] items and keeps track of
/// how many of them are fatal errors. This allows the compiler to collect multiple
/// errors before aborting, improving the development experience by reporting more
/// issues at once.
#[derive(Debug)]
pub struct Diagnostics {
    inner: Vec<TypeCheckDiagnostic>,
    fatal: usize,
}

impl Diagnostics {
    /// Create a new, empty diagnostics collection.
    #[must_use]
    pub const fn new() -> Self {
        Self {
            inner: Vec::new(),
            fatal: 0,
        }
    }

    pub fn clear(&mut self) {
        self.inner.clear();
        self.fatal = 0;
    }

    /// Returns the number of fatal errors in the diagnostics collection.
    #[must_use]
    pub const fn fatal(&self) -> usize {
        self.fatal
    }

    /// Returns the number of diagnostics in the collection.
    #[must_use]
    pub const fn len(&self) -> usize {
        self.inner.len()
    }

    /// Returns `true` if the diagnostics collection is empty.
    #[must_use]
    pub const fn is_empty(&self) -> bool {
        self.inner.is_empty()
    }

    /// Adds a new diagnostic to the collection.
    pub(crate) fn push(&mut self, diagnostic: TypeCheckDiagnostic) {
        if diagnostic.severity.is_fatal() {
            self.fatal += 1;
        }

        self.inner.push(diagnostic);
    }

    /// Merges another diagnostics collection into this one.
    ///
    /// This adds the diagnostics from the other collection to this one and updates the fatal error
    /// count accordingly.
    pub fn merge(&mut self, other: Self) {
        self.fatal += other.fatal;
        self.inner.extend(other.inner);
    }

    /// Consumes the diagnostics collection and returns a vector of diagnostics.
    #[must_use]
    pub fn into_vec(self) -> Vec<TypeCheckDiagnostic> {
        self.inner
    }
}

impl Default for Diagnostics {
    fn default() -> Self {
        Self::new()
    }
}
