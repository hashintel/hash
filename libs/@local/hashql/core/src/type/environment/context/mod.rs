pub mod diagnostics;
pub(crate) mod provision;
pub(crate) mod replace;
pub mod variance;
