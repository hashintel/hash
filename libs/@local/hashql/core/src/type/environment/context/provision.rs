use alloc::rc::Rc;
use core::cell::RefCell;

use crate::{collection::FastHashMap, id::Id, intern::Provisioned};

pub struct ProvisionedGuard<T: Id> {
    inner: Rc<ProvisionedScope<T>>,

    id: T,
    provisioned: Provisioned<T>,
    previous: Option<Provisioned<T>>,
}

impl<T> Drop for ProvisionedGuard<T>
where
    T: Id,
{
    fn drop(&mut self) {
        self.inner.exit(self);
    }
}

/// Manages provisional type mappings during the type simplification process.
///
/// This struct holds the forward (`T` -> `Provisioned<T>`) and reverse
/// (`Provisioned<T>` -> `T`) mappings used to track types that have been
/// temporarily 'provisioned' (allocated a placeholder ID) before their final
/// simplified form is determined and interned.
///
/// It is shared via `Rc` between the `SimplifyEnvironment` and active `ProvisionGuard`
/// instances, enabling RAII cleanup of provisional mappings when guards go out of scope.
///
/// ## Implementation Details: `RefCell` Usage
///
/// The `forward` and `reverse` maps are wrapped in `RefCell` for the following reasons:
///
/// 1. **Shared Ownership (`Rc`)**: The `Provision` struct itself is wrapped in an `Rc`. This allows
///    shared ownership between the `SimplifyEnvironment` and multiple active `ProvisionGuard`
///    instances.
///
/// 2. **Interior Mutability for Guards**: `ProvisionGuard` instances need to modify the `forward`
///    and `reverse` maps within their `Drop` implementation for RAII cleanup (removing the
///    provisional mappings they created). Since the `Provision` data is shared via `Rc`, direct
///    mutable borrowing (`&mut`) isn't possible from the guards. `RefCell` provides the necessary
///    interior mutability, allowing `borrow_mut()` within the guard's `drop` method (specifically,
///    in `Provision::exit`).
///
/// 3. **Enabling Mutation via Shared Reference (`Rc`)**: The core challenge is that
///    `ProvisionGuard` needs to mutate the `Provision` state during `drop`, but it only holds an
///    `Rc<Provision>` (a shared reference). `RefCell` provides the mechanism (`borrow_mut()`) to
///    achieve this *interior mutability*. The safety of these mutations, and reads like
///    `substitute`, (i.e., preventing conflicting borrows at runtime) is guaranteed because all
///    operations accessing the `Provision` struct occur through `SimplifyEnvironment` methods that
///    require `&mut self`. This `&mut self` ensures exclusive access to the `SimplifyEnvironment`
///    (and thus indirectly to the `Provision` maps) within the single thread at any given time.
///
/// 4. **Single-Threaded Context**: This approach relies on the assumption of a single-threaded
///    execution environment (consistent with `Heap` not being thread-safe). `RefCell` is suitable
///    here; a `Mutex` would be unnecessary overhead.
///
/// **In Summary**: `RefCell` is used here primarily because `Provision` is shared via `Rc`. It
/// provides the necessary *interior mutability* for `ProvisionGuard`'s RAII cleanup
/// (`Drop` impl) to modify the shared state through the `Rc`. The exclusive `&mut self`
/// required by the public methods of `SimplifyEnvironment` that interact with `Provision`
/// guarantees that there are no concurrent accesses within the same thread, making the use
/// of `RefCell` safe without needing a `Mutex`.
#[derive(Debug)]
pub(crate) struct ProvisionedScope<T> {
    forward: RefCell<FastHashMap<T, Provisioned<T>>>,
    reverse: RefCell<FastHashMap<Provisioned<T>, T>>,
}

impl<T> ProvisionedScope<T>
where
    T: Id,
{
    pub(crate) fn enter(self: Rc<Self>, id: T, provisioned: Provisioned<T>) -> ProvisionedGuard<T> {
        let previous = { self.forward.borrow_mut().insert(id, provisioned) };

        self.reverse.borrow_mut().insert(provisioned, id);

        ProvisionedGuard {
            inner: Rc::clone(&self),
            id,
            provisioned,
            previous,
        }
    }

    fn exit(&self, guard: &ProvisionedGuard<T>) {
        if let Some(previous) = guard.previous {
            self.forward.borrow_mut().insert(guard.id, previous);
        } else {
            self.forward.borrow_mut().remove(&guard.id);
        }

        self.reverse.borrow_mut().remove(&guard.provisioned);
    }

    pub(crate) fn get_substitution(&self, id: T) -> Option<T> {
        self.forward
            .borrow()
            .get(&id)
            .map(|provisioned| provisioned.value())
    }

    pub(crate) fn get_source(&self, id: T) -> Option<T> {
        self.reverse.borrow().get(&id).copied()
    }
}

impl<T> Default for ProvisionedScope<T> {
    fn default() -> Self {
        Self {
            forward: RefCell::default(),
            reverse: RefCell::default(),
        }
    }
}
