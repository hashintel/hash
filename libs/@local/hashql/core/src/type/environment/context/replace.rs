use alloc::rc::Rc;
use core::{cell::RefCell, hash::Hash};

use crate::collection::FastHashMap;

pub struct ReplacementGuard<T: Copy + Eq + Hash> {
    inner: Rc<ReplacementScope<T>>,

    items: Vec<(T, Option<T>)>,
}

impl<T> Drop for ReplacementGuard<T>
where
    T: Copy + Eq + Hash,
{
    fn drop(&mut self) {
        self.inner.exit(&mut self.items);
    }
}

#[derive(Debug)]
pub(crate) struct ReplacementScope<T> {
    // For the motivation of using RefCell, please refer to the documentation on
    // `ProvisionedScope`.
    lookup: RefCell<FastHashMap<T, T>>,
}

impl<T> ReplacementScope<T>
where
    T: Copy + Eq + Hash,
{
    pub(crate) fn enter_many(self: Rc<Self>, items: Vec<(T, T)>) -> ReplacementGuard<T> {
        let mut lookup = self.lookup.borrow_mut();
        let items = items
            .into_iter()
            .map(|(original, replacement)| {
                let previous = lookup.insert(original, replacement);
                (original, previous)
            })
            .collect();
        drop(lookup);

        ReplacementGuard { inner: self, items }
    }

    pub(crate) fn exit(&self, items: &mut Vec<(T, Option<T>)>) {
        if items.is_empty() {
            return;
        }

        let mut lookup = self.lookup.borrow_mut();

        for (original, previous) in items.drain(..) {
            if let Some(previous) = previous {
                lookup.insert(original, previous);
            } else {
                lookup.remove(&original);
            }
        }
    }

    pub(crate) fn lookup(&self, id: T) -> Option<T> {
        self.lookup.borrow().get(&id).copied()
    }
}

impl<T> Default for ReplacementScope<T> {
    fn default() -> Self {
        Self {
            lookup: RefCell::default(),
        }
    }
}
