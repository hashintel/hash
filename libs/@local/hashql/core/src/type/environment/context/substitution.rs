use hashbrown::HashMap;

use crate::r#type::{TypeId, kind::generic_argument::GenericArgumentId};

// The result post inference step (to be implemented)
#[derive(Debug)]
pub struct Substitution {
    arguments: HashMap<GenericArgumentId, TypeId, foldhash::fast::RandomState>,
    inference: HashMap<TypeId, TypeId, foldhash::fast::RandomState>,
}

impl Substitution {
    #[must_use]
    pub fn new() -> Self {
        Self {
            arguments: HashMap::default(),
            inference: HashMap::default(),
        }
    }

    #[must_use]
    pub fn argument(&self, id: GenericArgumentId) -> Option<TypeId> {
        self.arguments.get(&id).copied()
    }

    #[must_use]
    pub fn infer(&self, id: TypeId) -> Option<TypeId> {
        self.inference.get(&id).copied()
    }
}

impl Default for Substitution {
    fn default() -> Self {
        Self::new()
    }
}
