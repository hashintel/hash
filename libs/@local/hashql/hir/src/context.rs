use hashql_core::{heap::Heap, id::IdCounter, module::ModuleRegistry, symbol::SymbolTable};

use crate::{
    intern::Interner,
    map::HirMap,
    node::{HirId, r#let::VarId},
};

pub type BinderSymbolTable<'heap> = SymbolTable<'heap, VarId>;

#[derive(Debug)]
pub struct SymbolRegistry<'heap> {
    pub binder: BinderSymbolTable<'heap>,
}

impl SymbolRegistry<'_> {
    #[must_use]
    pub const fn new() -> Self {
        Self {
            binder: BinderSymbolTable::dense(),
        }
    }
}

impl Default for SymbolRegistry<'_> {
    fn default() -> Self {
        Self::new()
    }
}

#[derive(Debug)]
pub struct Counter {
    pub var: IdCounter<VarId>,
    pub hir: IdCounter<HirId>,
}

impl Counter {
    const fn new() -> Self {
        Self {
            var: IdCounter::new(),
            hir: IdCounter::new(),
        }
    }
}

impl Default for Counter {
    fn default() -> Self {
        Self::new()
    }
}

#[derive(Debug)]
pub struct HirContext<'env, 'heap> {
    pub heap: &'heap Heap,
    pub symbols: SymbolRegistry<'heap>,
    pub interner: &'env Interner<'heap>,
    pub modules: &'env ModuleRegistry<'heap>,
    pub counter: Counter,
    pub map: HirMap<'heap>,
}

impl<'env, 'heap> HirContext<'env, 'heap> {
    pub fn new(interner: &'env Interner<'heap>, modules: &'env ModuleRegistry<'heap>) -> Self {
        Self {
            heap: modules.heap,
            symbols: SymbolRegistry::new(),
            interner,
            modules,
            counter: Counter::new(),
            map: HirMap::new(),
        }
    }
}
