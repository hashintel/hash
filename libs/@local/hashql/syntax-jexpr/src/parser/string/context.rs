use core::ops::Range;

use hashql_ast::heap::Heap;
use hashql_core::span::{SpanId, TextRange, TextSize, storage::SpanStorage};
use winnow::{LocatingSlice, Stateful};

use crate::span::Span;

#[expect(
    clippy::cast_possible_truncation,
    reason = "The lexer ensures that we never have more than 4GiB of text"
)]
fn range_to_text(range: Range<usize>) -> TextRange {
    TextRange::new(
        TextSize::from((range.start + 1) as u32),
        TextSize::from((range.end + 1) as u32),
    )
}

#[derive(Debug, Copy, Clone)]
pub(crate) struct Context<'heap, 'span> {
    pub heap: &'heap Heap,
    pub spans: &'span SpanStorage<Span>,
    pub parent: SpanId,
}

impl Context<'_, '_> {
    pub(crate) fn span(&self, range: Range<usize>) -> SpanId {
        // `+ 1` here to offset the opening quote
        self.spans.insert(Span {
            range: range_to_text(range),
            pointer: None,
            parent_id: Some(self.parent),
        })
    }

    pub(crate) fn cover(&self, span: SpanId, range: Range<usize>) {
        self.spans.modify(span, |span| {
            span.range = span.range.cover(range_to_text(range.clone()));
        });
    }
}

pub(crate) type Input<'heap, 'span, 'source> =
    Stateful<LocatingSlice<&'source str>, Context<'heap, 'span>>;
