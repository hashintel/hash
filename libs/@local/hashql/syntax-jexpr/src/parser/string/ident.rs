use ada_url::{SchemeType, Url};
use hashql_core::{
    heap::Heap,
    symbol::{Ident, IdentKind, Symbol},
};
use unicode_normalization::{IsNormalized, UnicodeNormalization as _, is_nfc_quick};
use unicode_properties::{GeneralCategoryGroup, UnicodeGeneralCategory as _};
use winnow::{
    ModalResult, Parser as _,
    combinator::{alt, cut_err, delimited, dispatch, fail, peek, preceded},
    error::{AddContext, ParserError, StrContext, StrContextValue},
    token::{any, one_of, take_while},
};

use super::context::Input;

fn intern<'heap>(heap: &'heap Heap, value: &str) -> Symbol<'heap> {
    if is_nfc_quick(value.chars()) == IsNormalized::Yes {
        heap.intern_symbol(value)
    } else {
        let normalized: String = value.nfc().collect();
        heap.intern_symbol(&normalized)
    }
}

fn parse_ident_lexical<'heap, 'span, 'source, E>(
    input: &mut Input<'heap, 'span, 'source>,
) -> ModalResult<Ident<'heap>, E>
where
    E: ParserError<Input<'heap, 'span, 'source>>
        + AddContext<Input<'heap, 'span, 'source>, StrContext>,
{
    let context = input.state;

    let used = (
        one_of(unicode_ident::is_xid_start),
        cut_err(take_while(0.., unicode_ident::is_xid_continue)),
    );

    let unused = (
        '_',
        cut_err(take_while(1.., unicode_ident::is_xid_continue)),
    );

    alt((used, unused))
        .take()
        .with_span()
        .map(|(value, span): (&str, _)| Ident {
            span: context.span(span),
            value: intern(context.heap, value),
            kind: IdentKind::Lexical,
        })
        .parse_next(input)
}

fn is_symbol(char: char) -> bool {
    match char.as_ascii().map(|char| char as u8) {
        Some(
            b'!' | b'#' | b'$' | b'%' | b'&' | b'*' | b'+' | b'.' | b'/' | b'<' | b'=' | b'>'
            | b'?' | b'@' | b'\\' | b'^' | b'|' | b'-' | b'~',
        ) => true,
        Some(_) => false,
        None => {
            matches!(
                char.general_category_group(),
                GeneralCategoryGroup::Punctuation | GeneralCategoryGroup::Symbol
            )
        }
    }
}

fn parse_ident_symbol<'heap, 'span, 'source, E>(
    input: &mut Input<'heap, 'span, 'source>,
) -> ModalResult<Ident<'heap>, E>
where
    E: ParserError<Input<'heap, 'span, 'source>>
        + AddContext<Input<'heap, 'span, 'source>, StrContext>,
{
    let context = input.state;

    let bare = || take_while(1.., is_symbol);
    let escaped = delimited(
        '`',
        bare(),
        cut_err('`').context(StrContext::Expected(StrContextValue::CharLiteral('`'))),
    );

    alt((escaped, bare()))
        .with_span()
        .map(|(value, span): (&str, _)| Ident {
            span: context.span(span),
            value: intern(context.heap, value),
            kind: IdentKind::Symbol,
        })
        .parse_next(input)
}

// see: https://www.ietf.org/rfc/rfc3986.txt
static ALLOWED_URL_CHARS: [bool; 256] = {
    const RS: bool = true; // reserved
    const UR: bool = true; // unreserved
    const PC: bool = true; // percent
    const __: bool = false; // not allowed

    [
        //   1   2   3   4   5   6   7   8   9   A   B   C   D   E   F
        __, __, __, __, __, __, __, __, __, __, __, __, __, __, __, __, // 0
        __, __, __, __, __, __, __, __, __, __, __, __, __, __, __, __, // 1
        __, RS, __, RS, RS, PC, RS, RS, RS, RS, RS, RS, RS, UR, UR, RS, // 2
        UR, UR, UR, UR, UR, UR, UR, UR, UR, UR, RS, RS, __, RS, __, RS, // 3
        RS, UR, UR, UR, UR, UR, UR, UR, UR, UR, UR, UR, UR, UR, UR, UR, // 4
        UR, UR, UR, UR, UR, UR, UR, UR, UR, UR, UR, RS, __, RS, __, UR, // 5
        __, UR, UR, UR, UR, UR, UR, UR, UR, UR, UR, UR, UR, UR, UR, UR, // 6
        UR, UR, UR, UR, UR, UR, UR, UR, UR, UR, UR, __, __, __, UR, __, // 7
        __, __, __, __, __, __, __, __, __, __, __, __, __, __, __, __, // 8
        __, __, __, __, __, __, __, __, __, __, __, __, __, __, __, __, // 9
        __, __, __, __, __, __, __, __, __, __, __, __, __, __, __, __, // A
        __, __, __, __, __, __, __, __, __, __, __, __, __, __, __, __, // B
        __, __, __, __, __, __, __, __, __, __, __, __, __, __, __, __, // C
        __, __, __, __, __, __, __, __, __, __, __, __, __, __, __, __, // D
        __, __, __, __, __, __, __, __, __, __, __, __, __, __, __, __, // E
        __, __, __, __, __, __, __, __, __, __, __, __, __, __, __, __, // F
    ]
};

fn is_url_char(char: char) -> bool {
    char.as_ascii()
        .is_some_and(|char| ALLOWED_URL_CHARS[char as usize])
}

fn parse_ident_url<'heap, 'span, 'source, E>(
    input: &mut Input<'heap, 'span, 'source>,
) -> ModalResult<Ident<'heap>, E>
where
    E: ParserError<Input<'heap, 'span, 'source>>
        + AddContext<Input<'heap, 'span, 'source>, StrContext>,
{
    let context = input.state;

    let url = (one_of(is_url_char), cut_err(take_while(1.., is_url_char)))
        .take()
        .verify(|value: &str| {
            // check if the value is an actual URL, we have only parsed what *looks* like a URL
            let Ok(url) = Url::parse(value, None) else {
                return false;
            };

            let scheme = matches!(url.scheme_type(), SchemeType::Http | SchemeType::Https);
            let ends_with_slash = value.ends_with('/');

            scheme && ends_with_slash
        })
        .context(StrContext::Expected(StrContextValue::Description(
            "http(s) url with trailing `/`",
        )));

    delimited(
        '`',
        url,
        cut_err('`').context(StrContext::Expected(StrContextValue::CharLiteral('`'))),
    )
    .with_span()
    .map(|(url, span)| Ident {
        span: context.span(span),
        value: intern(context.heap, url),
        kind: IdentKind::BaseUrl,
    })
    .parse_next(input)
}

pub(crate) fn parse_ident<'heap, 'span, 'source, E>(
    input: &mut Input<'heap, 'span, 'source>,
) -> ModalResult<Ident<'heap>, E>
where
    E: ParserError<Input<'heap, 'span, 'source>>
        + AddContext<Input<'heap, 'span, 'source>, StrContext>,
{
    dispatch! {peek(any);
        '`' => cut_err(alt((parse_ident_symbol, parse_ident_url))),
        char if unicode_ident::is_xid_start(char) || char == '_' => cut_err(parse_ident_lexical),
        char if is_symbol(char) => cut_err(parse_ident_symbol),
        _ => fail
    }
    .context(StrContext::Label("identifier"))
    .parse_next(input)
}

pub(crate) fn parse_ident_labelled_argument<'heap, 'span, 'source, E>(
    input: &mut Input<'heap, 'span, 'source>,
) -> ModalResult<Ident<'heap>, E>
where
    E: ParserError<Input<'heap, 'span, 'source>>
        + AddContext<Input<'heap, 'span, 'source>, StrContext>,
{
    preceded(
        cut_err(':').context(StrContext::Expected(StrContextValue::CharLiteral(':'))),
        parse_ident,
    )
    .parse_next(input)
}

#[cfg(test)]
mod tests {
    #![expect(clippy::non_ascii_literal)]

    use super::parse_ident;
    use crate::parser::string::test::{bind_parser, test_cases};

    bind_parser!(Debug; fn parse(parse_ident));

    // Test cases for lexical identifiers
    test_cases! {parse;
        lexical_basic("hello") => "Basic lexical identifier",
        lexical_with_underscore("_hello") => "Lexical identifier starting with underscore",
        lexical_with_numbers("hello123") => "Lexical identifier with numbers",
        lexical_mixed_case("HelloWorld") => "Lexical identifier with mixed case",
        lexical_with_trailing("hello_world after") => "Lexical identifier with trailing content",
    }

    // Test cases for Unicode identifiers
    test_cases! {parse;
        unicode_japanese("こんにちは") => "Japanese Unicode identifier",
        unicode_mixed("hello世界") => "Mixed ASCII and Unicode identifier",
        unicode_with_emoji("x☺") => "Identifier with emoji",
        unicode_precomposed("café") => "Identifier with precomposed character",
        unicode_decomposed("cafe\u{0301}") => "Identifier with decomposed character",
    }

    // Test cases for symbol identifiers
    test_cases! {parse;
        symbol_basic("++") => "Basic symbol",
        symbol_complex("<=>") => "Complex symbol combination",
        symbol_escaped("`++`") => "Escaped symbol",
        symbol_escaped_with_trailing("`++` next") => "Escaped symbol with trailing content",
        symbol_unicode("…") => "Unicode symbol",
        symbol_unicode_complex("≤≥≠") => "Complex Unicode symbols",
        symbol_unicode_escaped("`§¶†‡`") => "Escaped Unicode symbols",
    }

    // Test cases for URL identifiers
    test_cases! {parse;
        url_http("`http://example.com/`") => "Basic HTTP URL",
        url_https("`https://example.com/`") => "HTTPS URL",
        url_with_path("`https://example.com/path/to/resource/`") => "URL with path",
        url_with_query("`https://example.com/?query=value&param=123/`") => "URL with query parameters",
        url_with_trailing("`https://example.com/` next") => "URL with trailing content",
        url_with_port("`https://example.com:8080/`") => "URL with port",
        url_with_auth("`https://user:pass@example.com/`") => "URL with authentication",
        url_with_fragment("`https://example.com/#fragment/`") => "URL with fragment",
    }

    // Test cases for invalid inputs
    test_cases! {parse;
        invalid_url_no_trailing_slash("`https://example.com`") => "URL without trailing slash (invalid)",
        invalid_url_not_http("`ftp://example.com/`") => "Non-HTTP/HTTPS URL (invalid)",
        invalid_url_malformed("`http:///invalid/`") => "Malformed URL (invalid)",
        invalid_unclosed_backtick("`unclosed") => "Unclosed backtick (invalid)",
        invalid_unclosed_backtick_url("`https://example.com/") => "Unclosed backtick in URL (invalid)",
        invalid_empty("") => "Empty input (invalid)",
        invalid_whitespace("   ") => "Whitespace only (invalid)",
    }

    // Test for colon in symbols - should be rejected
    test_cases! {parse;
        invalid_colon_symbol(":") => "Colon as bare symbol (should be rejected)",
        invalid_colon_symbol_compound("::") => "Compound colon symbol (should be rejected)",
        invalid_colon_symbol_mixed("->:") => "Symbol with colon (should be rejected)",
        invalid_colon_symbol_assign(":=") => "Assignment with colon (should be rejected)",
        invalid_colon_symbol_backticks("`:`") => "Colon in backticks (should be rejected)",
        invalid_colon_symbol_backticks_compound("`:=`") => "Compound colon in backticks (should be rejected)",
        invalid_colon_in_fake_url("`example:not-url/`") => "Colon in invalid URL-like string (should be rejected)",
    }

    // Test mixed input scenarios
    test_cases! {parse;
        mixed_lexical_then_symbol("hello++") => "Lexical followed by symbol without space",
        mixed_with_whitespace("hello `symbol`") => "Lexical then symbol with space",
    }

    // Additional test for specific edge cases
    test_cases! {parse;
        edge_single_underscore("_") => "Single underscore identifier",
        edge_single_character("x") => "Single character identifier",
        edge_all_numbers_invalid("123") => "All numbers (invalid as lexical identifier)",
        edge_dash_only("-") => "Single dash as symbol",
        edge_complex_operators("<=>=><=>") => "Complex operator chain",
    }

    // Unicode normalization and edge cases
    test_cases! {parse;
        unicode_combining_marks("a\u{0308}\u{0323}") => "Identifier with multiple combining marks",
        unicode_homoglyphs("рaypal") => "Identifier with Cyrillic 'р' instead of Latin 'p'",
        unicode_zero_width("hello\u{200B}world") => "Identifier with zero-width space",
        unicode_normalization_nfd("a\u{0308}") => "Character with combining diaeresis (NFD)",
    }

    // Complex symbol tests
    test_cases! {parse;
        symbol_very_long(&"*".repeat(100)) => "Very long repeated symbol",
        symbol_mixed_blocks("≈∞♥★") => "Symbols from different Unicode blocks",
        symbol_with_whitespace("`+ +`") => "Symbol with internal whitespace in backticks",
    }

    // Complex URL tests
    test_cases! {parse;
        url_idn("`https://例子.测试/`") => "URL with internationalized domain name",
        url_ipv6("`https://[2001:db8::1]/`") => "URL with IPv6 address",
        url_percent_encoded("`https://example.com/%E2%82%AC/`") => "URL with percent-encoded characters",
        url_complex("`https://sub.many.levels.example.co.uk:8443/very/deep/path/?q=complex&t=true#section/`") => "Complex URL with multiple components",
    }

    // Error cases and boundaries
    test_cases! {parse;
        error_valid_start_invalid_continue("a\u{0000}bc") => "Identifier with null character",
        error_max_length_exceeded(&"a".repeat(10000)) => "Extremely long identifier",
        error_control_chars("test\u{0007}beep") => "Identifier with control character",
    }

    #[test]
    fn unicode_normalization() {
        // The precomposed character "é" (U+00E9)
        let precomposed = "café";

        // The decomposed version: "e" + combining acute accent (U+0065 U+0301)
        let decomposed = "cafe\u{0301}";

        // Parse both forms
        let (precomposed_result, _) = parse(precomposed);
        let (decomposed_result, _) = parse(decomposed);

        // They should be identical after parsing due to normalization
        assert_eq!(precomposed_result, decomposed_result);
    }
}
