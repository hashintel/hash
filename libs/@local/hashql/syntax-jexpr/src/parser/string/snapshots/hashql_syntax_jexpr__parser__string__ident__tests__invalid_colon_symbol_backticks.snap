---
source: libs/@local/hashql/syntax-jexpr/src/parser/string/ident.rs
description: Colon in backticks (should be rejected)
expression: "`:`"
info:
  kind: Err
---
`:`
  ^
invalid identifier
expected http(s) url with trailing `/`
